mmtk/src/gc_trigger.rs

```diff
@@ -0,0 +1,330 @@
+                target_allocs = max_target_allocs as f64;
+                info!(
+                    "Target allocs exceeds rate limit max, target_allocs = {}",
+                    target_allocs
+                );
+            }
+
+            if target_allocs < min_target_allocs as f64 {
+                target_allocs = min_target_allocs as f64;
+                info!(
+                    "Target allocs below rate limit min, target_allocs = {}",
+                    target_allocs
+                );
+            }
+
+            target_heap = target_allocs + heap_size as f64;
+            if target_heap < DEFAULT_COLLECT_INTERVAL as f64 {
+                target_heap = DEFAULT_COLLECT_INTERVAL as f64;
+                info!(
+                    "Target heap set to default interval, target_heap = {}",
+                    target_heap
+                );
+            }
+            self.heap_target
+                .store(target_heap as usize, Ordering::Relaxed);
+        }
+    }
+
+    fn on_pending_allocation(&self, pages: usize) {
+        self.pending_pages.fetch_add(pages, Ordering::SeqCst);
+    }
+
+    /// Is a GC required now?
+    fn is_gc_required(
+        &self,
+        space_full: bool,
+        space: Option<SpaceStats<JuliaVM>>,
+        plan: &dyn Plan<VM = JuliaVM>,
+    ) -> bool {
+        if self.is_heap_full(plan) {
+            return true;
+        }
+
+        plan.collection_required(space_full, space)
+    }
+
+    /// Is current heap full?
+    fn is_heap_full(&self, plan: &dyn Plan<VM = JuliaVM>) -> bool {
+        // reserved pages now should include pending allocations
+        let reserved_pages_now =
+            plan.get_reserved_pages() + self.pending_pages.load(Ordering::SeqCst);
+
+        let heap_size = conversions::pages_to_bytes(reserved_pages_now);
+
+        trace!(
+            "Heap size = {}, heap target = {}",
+            heap_size,
+            self.heap_target.load(Ordering::Relaxed)
+        );
+
+        heap_size >= self.heap_target.load(Ordering::Relaxed)
+    }
+
+    /// Return the current heap size (in pages)
+    fn get_current_heap_size_in_pages(&self) -> usize {
+        self.heap_target.load(Ordering::Relaxed) / BYTES_IN_PAGE
+    }
+
+    /// Return the upper bound of heap size
+    fn get_max_heap_size_in_pages(&self) -> usize {
+        self.max_total_memory.load(Ordering::Relaxed) / BYTES_IN_PAGE
+    }
+
+    /// Can the heap size grow?
+    fn can_heap_size_grow(&self) -> bool {
+        true
+    }
+}
+
+// copy of jl_gc_smooth from gc-stock.c
+fn mmtk_jl_gc_smooth(old_val: usize, new_val: usize, factor: f64) -> usize {
+    let est = factor * old_val as f64 + (1.0 - factor) * new_val as f64;
+    if est <= 1.0 {
+        1_usize
+    } else if est > (2usize << 36) as f64 {
+        2usize << 36
+    } else {
+        est as usize
+    }
+}
+
+fn mmtk_overallocation(old_val: usize, val: usize, max_val: usize) -> usize {
```

Comment:
Probably refer to the original function in Julia in the comments, and also copy the original comments here. The code looks quite confusing at the moment.